erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py

```diff
@@ -0,0 +1,241 @@
+			{"filters": filters, "data": res, "aging": aging[0] if doc.include_ageing else None})
+		html = frappe.render_template(base_template_path, {"body": html, "css": get_print_style(), "title": "Statement For " + entry.customer})
+		statement_dict[entry.customer] = html
+	if not bool(statement_dict):
+		return False
+	elif consolidated:
+		result = ''.join(list(statement_dict.values()))
+		return get_pdf(result, {'orientation': doc.orientation})
+	else:
+		for customer, statement_html in statement_dict.items():
+			statement_dict[customer]=get_pdf(statement_html, {'orientation': doc.orientation})
+		return statement_dict
+
+@frappe.whitelist()
+def fetch_customers(customer_collection, collection_name, primary_mandatory):
+	print(type(primary_mandatory))
+	customer_list = []
+	customers = []
+	if customer_collection == 'Sales Person':
+		customer_list = get_customers_based_on_sales_person(collection_name)
+
+		if not bool(customer_list):
+			frappe.throw('No Customers found with selected options!')
+
+		customers = frappe.get_list('Customer', fields=['name', 'email_id'], \
+			filters=[['name', 'in', list(customer_list['Customer'])]])
+		customer_list = []
+	else:
+		field_dict = {
+			'Customer Group': 'customer_group',
+			'Territory': 'territory',
+			'Sales Partner': 'default_sales_partner'
+		}
+		if customer_collection == 'Sales Partner':
+			customers = frappe.get_list('Customer', fields=['name', 'email_id'], \
+				filters=[[field_dict[customer_collection], '=', collection_name]])
+		else:
+			collection = frappe.get_doc(customer_collection, collection_name)
+			selected = frappe.get_list(customer_collection, filters=[
+					['lft', '>=', collection.lft],
+					['rgt', '<=', collection.rgt]
+				],
+				fields=['name'],
+				order_by='lft asc, rgt desc'
+			)
+			selected = [customer.name for customer in selected]
+			customers = frappe.get_list('Customer', fields=['name', 'email_id'], \
+				filters=[[field_dict[customer_collection], 'IN', selected]])
+	for customer in customers:
+		primary_email = customer.get('email_id') or ''
+		billing_email = get_customer_emails(customer.name, 1, billing_and_primary=False)
+
+		if billing_email == '' or (primary_email == '' and int(primary_mandatory)):
+			continue
+
+		customer_list.append({
+			'name': customer.name,
+			'primary_email': primary_email,
+			'billing_email': billing_email
+		})
+	return customer_list
+
+@frappe.whitelist()
+def get_customer_emails(customer_name, primary_mandatory, billing_and_primary=True):
+	billing_email = frappe.db.sql("""
+		SELECT c.email_id FROM `tabContact` AS c JOIN `tabDynamic Link` AS l ON c.name=l.parent \
+		WHERE l.link_doctype='Customer' and l.link_name='""" + customer_name + """' and \
+		c.is_billing_contact=1 \
+		order by c.creation desc""")
+
+	if len(billing_email) == 0 or (billing_email[0][0] is None):
+		if billing_and_primary:
+			frappe.throw('No billing email found for customer: '+ customer_name)
```

Comment:
If no billing email, but there is a primary email, don't throw error, send the email to primary contact.